consumer.go

```diff
@@ -514,7 +514,7 @@ func (child *partitionConsumer) parseMessages(msgSet *MessageSet) ([]*ConsumerMe
 		}
 	}
 
-	if incomplete || len(messages) == 0 {
+	if incomplete && len(messages) == 0 {
```

Comment:
I don't understand this change; `incomplete` can only be set if `messages` contains at least one complete message already, so this condition becomes impossible to satisfy? I assume this is the `kafka: response did not contain all the expected topic/partition blocks` fix you mention; I'd appreciate more details on that.

I'm looking at this method again, and I'm wondering if the `incomplete` boolean is even necessary anymore. I'm not sure what it was trying to accomplish originally (guarding against decreasing offsets maybe?), but a simple check on `len(messages) == 0` seems to be sufficient on its own? I'm not sure. That might even mean we don't need `prelude` anymore either, and can just unconditionally skip messages with `offset < child.offset`.

Reply:
You right, the error got away because this condition is now impossible to achieve. My reasoning was that incomplete record is supposed to be claimed as error only if there are no complete messages in the batch/set. But now looking at incomplete I understand that it actually does not mean partial trailing message as I thought.

I do not think that event `len(message) == 0` should be treated as an error. If Kafka broker decides to return an empty block, so what? it just means that there are no new message and sarama should just make another long polling fetch request.

Reply:
Ah, so the case where there are legitimately no messages in the block is already handled in `parseResponse`. If we get here and there are no messages we could extract, then I think it's legitimate to conclude an error of some sort since the server presumably sent us only messages we didn't ask for? I don't know if it's OK for the server to respond with a prelude but no actual messages...

Reply:
I think the reason why I saw `kafka: response did not contain all the expected topic/partition blocks` error while running this test, was exactly because the server was replying with a sets/batches containing prelude only. I will double check that tomorrow when I am back in the office.